staker-rewards/lib/conditions.js

```diff
@@ -0,0 +1,180 @@
+import { callWithRetry } from "./contract-helper.js"
+
+export default class Conditions {
+  constructor(
+    context,
+    interval,
+    bondedECDSAKeepFactory,
+    keepBonding,
+    sortitionPoolAddress
+  ) {
+    this.context = context
+    this.interval = interval
+    this.bondedECDSAKeepFactory = bondedECDSAKeepFactory
+    this.keepBonding = keepBonding
+    this.sortitionPoolAddress = sortitionPoolAddress
+    this.operatorsDeauthorizedInInterval = []
+  }
+
+  static async initialize(context, interval) {
+    const { contracts } = context
+
+    const bondedECDSAKeepFactory = await contracts.BondedECDSAKeepFactory.deployed()
+
+    const sortitionPoolAddress = await callWithRetry(
+      bondedECDSAKeepFactory.methods.getSortitionPool(
+        contracts.sanctionedApplicationAddress
+      )
+    )
+
+    const conditions = new Conditions(
+      context,
+      interval,
+      bondedECDSAKeepFactory,
+      await contracts.KeepBonding.deployed(),
+      sortitionPoolAddress
+    )
+
+    return conditions
+  }
+
+  async checkDeauthorizations() {
+    // If Tenderly API is available refresh cached data.
+    if (this.context.tenderly) {
+      this.refreshDeauthorizationsCache()
+    }
+
+    // Fetch deauthorization transactions from cache.
+    // We expect only successful transactions to be returned, which means we don't
+    // need to double check authorizer vs operator as this was already handled
+    // by the contract on the function call.
+    const deauthorizations = this.context.cache.getFunctionCalls(
+      this.keepBonding.options.address,
+      "deauthorizeSortitionPoolContract"
+    )
+
+    console.debug(
+      `Found ${deauthorizations.length} sortition pool contract deauthorizations`
+    )
+
+    if (deauthorizations && deauthorizations.length > 0) {
+      for (let i = 0; i < deauthorizations.length; i++) {
+        const transaction = deauthorizations[i]
+
+        console.debug(`Checking transaction ${transaction.hash}`)
+
+        if (
+          transaction.block_number <= this.interval.startBlock ||
+          transaction.block_number >= this.interval.endBlock
```

Comment:
In other scripts we do `start <= interval < end` so that we do not have intervals overlapping on one block. For example, in `sla-calculator.js`:
```
    const isInInterval = (timestamp) =>
      interval.start <= timestamp && timestamp < interval.end
```

Reply:
Awesome point!

Reply:
Do we assume end block of an interval will be a starting block for the following interval? Won't we start a new interval with end block +1?

Reply:
Oh yeah, timestamps are start and end of the interval.